File: Programmers/BFS/가장 먼 노드.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

void BFS(int v, int n, vector<vector<int>>& graph, vector<int>& visited) {
    queue<int> q;
    q.push(v);
    visited[v] = 1;

    while(!q.empty()) {
        int cur = q.front();
        q.pop();

        for (int next : graph[cur]) {
            if (visited[next] == 0) {
                visited[next] = visited[cur] + 1;
                q.push(next);
            }
        }
    }
}

int solution(int n, vector<vector<int>> edge) {
    int answer = 0;
    vector<vector<int>> graph(n + 1, vector<int> ());
    vector<int> visited(n + 1, 0);

    for (auto tmp : edge) {
        graph[tmp[0]].push_back(tmp[1]);
        graph[tmp[1]].push_back(tmp[0]);
    }

    BFS(1, n, graph, visited);
    int maxLen = 0;
    for (int i = 1; i <= n; i++) {
        if (maxLen < visited[i]) maxLen = visited[i];
    }

    for (int i = 1; i <= n; i++) {
        if (maxLen == visited[i]) answer++;
    }

    return answer;
}

File: BOJ/BFS/가장 먼 노드.py
from collections import deque


def BFS(graph, visited, n, start):
    queue = deque()
    queue.append(start)

    while queue:
        cur = queue.popleft()
        for next in graph[cur]:
            if visited[next] == 0:
                visited[next] = visited[cur] + 1
                queue.append(next)


def solution(n, edge):
    answer = 0
    graph = [[] for i in range(n + 1)]
    visited = [0] * (n + 1)

    for u, v in edge:
        graph[u].append(v)
        graph[v].append(u)

    visited[1] = 1
    BFS(graph, visited, n, 1)

    max_dist = max(visited)
    answer = visited.count(max_dist)

    return answer